### PR TITLE
Enable Half/BFloat16 support for to_dense and coalesce methods.

### PR DESCRIPTION
**Pull request by pearu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #72633
* #71239
* #72398
* __->__ #72397

Differential Revision: [D34286114](https://our.internmc.facebook.com/intern/diff/D34286114)

### CODE REVIEW

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
Looks good! But just to be safe, could you also please audit some of the tests that exercise these functions and extend them with these new dtypes?

File: test/test_sparse.py [hunk 1]
```diff
@@ -3234,7 +3219,7 @@ def sparse_log(x):
                   *get_all_fp_dtypes(
                       include_half=(CUDA11OrLater and SM53OrLater),
                       include_bfloat16=(CUDA11OrLater and SM80OrLater)))
-    @precisionOverride({torch.bfloat16: 1e-2, torch.float16: 1e-2, torch.complex64: 1e-2, torch.float32: 1e-2})
+    @precisionOverride({torch.bfloat16: 5e-2, torch.float16: 5e-2, torch.complex64: 1e-2, torch.float32: 1e-2})
```
Comment (cpuhrsch) [inline test/test_sparse.py:3222]: Why is the decreased precision necessary? It seems an unrelated change.

File: test/test_sparse.py [hunk 2]
```diff
@@ -360,16 +354,9 @@ def fn(x):
             res = torch.empty((3, 4, 5, 0), dtype=dtype, device=device)
             test_tensor(x, res)
 
-    # half tensors on cpu don't implement to_dense, so need to convert to float
```
Comment (cpuhrsch) [inline test/test_sparse.py:363]: Awesome. Happy to see this removed.

**Comment by cpuhrsch** [comment]
Much better. See comments.

File: test/test_sparse.py (hunk 1 above)
Comment (pearu) [reply test/test_sparse.py:3222]: After enabling more tests in this PR, the seed to the random tensor generator changed leading to different inputs to this test function. The new inputs required a decrease of precision here (IIRC, the difference reported was about 2.2e-2).

File: test/test_sparse.py [hunk 3]
```diff
@@ -1991,8 +1978,7 @@ def is_integral(dtype):
                 sparse_tensor.requires_grad_()
 
     @coalescedonoff
-    @dtypes(*get_all_dtypes(include_bool=False, include_half=False,
-                            include_bfloat16=False, include_complex=False))
+    @dtypes(*get_all_dtypes(include_bool=False, include_half=False, include_complex=False))
```
[the remark this answers is not in the document]
Comment (pearu) [reply test/test_sparse.py:1981]: Yes, but it was not possible because of:
```
RuntimeError: "log1p" "_vml_cpu" not implemented for 'Half'
```

File: test/test_sparse.py (hunk 1 above)
Comment (pearu) [reply test/test_sparse.py:3222]: Ok, will do it.

Comment (pearu) [reply test/test_sparse.py]: Restored the original state as the tests pass again with the given precision parametes.

**Review by cpuhrsch — APPROVED** [review]

**Comment by cpuhrsch** [comment]
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34286114).
